<a href="https://colab.research.google.com/github/parekhravik/MLProjects/blob/master/1004_CapStoneProject_HR_ERIP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Team:** 1004
**Project Name:** Employee Retantion Insight Project (ERIP)

# 1. Problem Statement

Client is looking for an solution which can predict whether an employee will leave or stay based on the past experience data collected. It will give more proactive approach to HR team in handling attritions.


**Current Practice:**
* Once an employee leaves, he or she is taken an interview with the name “exit interview” and shares reasons for leaving. The HR Department then tries and learns insights from the interview and makes changes accordingly.

* This suffers from the following problems:

> 1. This approach is that it's too haphazard. The quality of insight gained from an interview depends heavily on the skill of the interviewer.
> 2. The second problem is these insights can't be aggregated and interlaced across all employees who have left.
> 3.  The third is that it is too late by the time the proposed policy changes take effect.

Available Data:

* Department Details
* Employee Personal Information (Non Sensitive)
* Employee Informations (Administrative, Work Load, Company Mutual Evaluation)

# 2. EDA


*   First Step will be to load the data and eximine.


In [0]:
!pip install mysql-connector-python

     |████████████████████████████████| 13.1MB 4.5MB/s 


In [0]:
#Imports

import mysql.connector
from mysql.connector import errorcode

import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib.pyplot import pie, axis, show
%matplotlib inline                                                              

import warnings                                                                 # Ignore warning related to pandas_profiling
warnings.filterwarnings('ignore') 

import sqlalchemy as sqla
import pandas_profiling

from google.colab import files

In [0]:
#Connect to DB and Load data

#DB Connection Details
vDBUser='student'
vDBPwd='student'
vDBHost='projects.insaid.co'
vDBSid='Capstone2'

try:
  cnx = mysql.connector.connect(user=vDBUser, password=vDBPwd,
                                host=vDBHost, database=vDBSid)
  
  pdDepartments = pd.read_sql('SELECT * FROM department_data',cnx) 
  pdEmployeeDetails = pd.read_sql('SELECT * FROM employee_details_data',cnx)  
  pdEmployeeData = pd.read_sql('SELECT * FROM employee_data',cnx)
  
except mysql.connector.Error as err:
  if err.errno == errorcode.ER_ACCESS_DENIED_ERROR:
    print("Something is wrong with your user name or password")
  elif err.errno == errorcode.ER_BAD_DB_ERROR:
    print("Database does not exist")
  else:
    print(err)
else:
  cnx.close()
  
cnx.close()

In [0]:
pdDepartments

dept_id    dept_name              dept_head
0    D00-IT           IT             Henry Adey
1    D00-SS        Sales        Edward J Bayley
2    D00-TP         Temp        Micheal Zachrey
3   D00-ENG  Engineering  Sushant Raghunathan K
4    D00-SP      Support         Amelia Westray
5    D00-FN      Finance              Aanchal J
6    D00-PR  Procurement            Louie Viles
7    D00-AD        Admin          Evelyn Tolson
8    D00-MN   Management          Ellie Trafton
9    D00-MT    Marketing           Reuben Swann
10   D00-PD      Product          Darcy Staines

In [0]:
pdDepartments.shape

(11, 3)

In [0]:
pdDepartments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11 entries, 0 to 10
Data columns (total 3 columns):
dept_id      11 non-null object
dept_name    11 non-null object
dept_head    11 non-null object
dtypes: object(3)
memory usage: 344.0+ bytes


# 2.1 Observations: department_data
1.   Our client company has total 11 departments. 
2.   This data has no missing values, each department is known by its dept_id.

Below are the columns in it:


1.   dept_id – Unique Department Code
2.   dept_name – Name of the Department
3.   dept_head – Name of the Head of the Department

In [0]:
pdEmployeeDetails.head()

employee_id  age  gender marital_status
0       113558   43    Male        Married
1       112256   24  Female      Unmarried
2       112586   22  Female      Unmarried
3       108071   36    Male        Married
4       116915   38    Male        Married

In [0]:
pdEmployeeDetails.shape

(14245, 4)

In [0]:
pdEmployeeDetails.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14245 entries, 0 to 14244
Data columns (total 4 columns):
employee_id       14245 non-null int64
age               14245 non-null int64
gender            14245 non-null object
marital_status    14245 non-null object
dtypes: int64(2), object(2)
memory usage: 445.2+ KB


In [0]:
pdEmployeeDetails.describe()

employee_id           age
count   14245.000000  14245.000000
mean   112123.050544     32.889926
std      8500.457343      9.970834
min    100101.000000     22.000000
25%    105775.000000     24.000000
50%    111298.000000     29.000000
75%    116658.000000     41.000000
max    148988.000000     57.000000

# 2.1 Observations: Employee_Details
Client has total 14245 Employee Details 

Below are the columns in it:
1. employee_id – Unique ID Number for each employee
2. age – Age of the employee
3. gender – Gender of the employee
4. marital_status – Marital Status of the employee

In [0]:
pdEmployeeData.head()

avg_monthly_hrs department  filed_complaint  ...  status  tenure  employee_id
0            246.0       None              NaN  ...    Left     4.0       124467
1            134.0       None              NaN  ...    Left     3.0       112210
2            156.0     D00-SS              1.0  ...    Left     3.0       126150
3            256.0     D00-SP              NaN  ...    Left     4.0       125346
4            146.0     D00-SS              NaN  ...    Left     3.0       113707

[5 rows x 11 columns]

In [0]:
pdEmployeeData.describe()

avg_monthly_hrs  filed_complaint  ...        tenure    employee_id
count     14150.000000           2046.0  ...  14000.000000   14150.000000
mean        199.994346              1.0  ...      3.499357  112080.750247
std          50.833697              0.0  ...      1.462584    8748.202856
min          49.000000              1.0  ...      2.000000       0.000000
25%         155.000000              1.0  ...      3.000000  105772.500000
50%         199.000000              1.0  ...      3.000000  111291.500000
75%         245.000000              1.0  ...      4.000000  116650.750000
max         310.000000              1.0  ...     10.000000  148988.000000

[8 rows x 8 columns]

In [0]:
pdEmployeeData.info()

In [0]:
pdEmployeeData.isnull().sum()

avg_monthly_hrs         0
department            707
filed_complaint         0
last_evaluation      1487
n_projects              0
recently_promoted       0
salary                  0
satisfaction          150
status                  0
tenure                150
employee_id             0
dtype: int64

In [0]:
pdEmployeeData["filed_complaint"].fillna(0, inplace=True)
pdEmployeeData["recently_promoted"].fillna(0, inplace=True)

In [0]:
pdEmployeeData.department.unique()

array([None, 'D00-SS', 'D00-SP', 'D00-MT', 'D00-PD', 'D00-IT', 'D00-AD',
       'D00-MN', 'D00-ENG', 'D00-PR', 'D00-TP', 'D00-FN', '-IT'],
      dtype=object)

In [0]:
pdEmployeeData["department"] = pdEmployeeData["department"].replace({'-IT':'D00-IT'}, regex=False)

In [0]:
pdEmployeeData["department"].isnull().sum()

707

In [0]:
mode = pdEmployeeData["department"].mode()
mode

0    D00-SS
dtype: object

In [0]:
pdEmployeeData["department"] = pdEmployeeData["department"].fillna(mode[0])
pdEmployeeData["department"].isnull().sum()

0

In [0]:
meanLE = pdEmployeeData["last_evaluation"].mean()
pdEmployeeData["last_evaluation"] = pdEmployeeData["last_evaluation"].fillna(meanLE)

In [0]:
meanS = pdEmployeeData["satisfaction"].mean()
pdEmployeeData["satisfaction"] = pdEmployeeData["satisfaction"].fillna(meanS)

In [0]:
meanT = pdEmployeeData["tenure"].mean()
pdEmployeeData["tenure"] = pdEmployeeData["tenure"].fillna(meanT)

In [0]:
pdEmployeeData.isnull().sum()

avg_monthly_hrs      0
department           0
filed_complaint      0
last_evaluation      0
n_projects           0
recently_promoted    0
salary               0
satisfaction         0
status               0
tenure               0
employee_id          0
dtype: int64

In [0]:
profile = pandas_profiling.ProfileReport(pdEmployeeData)
profile.to_file(outputfile="pdEmployeeData.html")

In [0]:
pdEmployeeData = pdEmployeeData[pdEmployeeData.employee_id != 0]

In [0]:
files.download("pdEmployeeData.html")

In [0]:
profile = pandas_profiling.ProfileReport(pdEmployeeData)
profile.to_file(outputfile="pdEmployeeData.html")

In [0]:
files.download("pdEmployeeData.html")

Let's join all 3 dataset and create a single one. Would be much better to get insights on all of the features togather.

In [0]:
pdEmployeeTemp = pd.merge(pdEmployeeData, pdEmployeeDetails, on = 'employee_id', how = 'left')

In [0]:
pdEmployeeTemp.head()

avg_monthly_hrs department  filed_complaint  ...   age  gender  marital_status
0            246.0       None              NaN  ...  42.0  Female         Married
1            134.0       None              NaN  ...  23.0  Female       Unmarried
2            156.0     D00-SS              1.0  ...  24.0  Female       Unmarried
3            256.0     D00-SP              NaN  ...  51.0  Female         Married
4            146.0     D00-SS              NaN  ...  23.0  Female       Unmarried

[5 rows x 14 columns]

In [0]:
pdEmployeeMerged = pd.merge(pdEmployeeTemp, pdDepartments, left_on='department',right_on='dept_id', how = 'left')

In [0]:
pdEmployeeMerged.head()

avg_monthly_hrs department  ...  dept_name        dept_head
0            246.0     D00-SS  ...      Sales  Edward J Bayley
1            134.0     D00-SS  ...      Sales  Edward J Bayley
2            156.0     D00-SS  ...      Sales  Edward J Bayley
3            256.0     D00-SP  ...    Support   Amelia Westray
4            146.0     D00-SS  ...      Sales  Edward J Bayley

[5 rows x 17 columns]

In [0]:
#Drop duplicate department id
pdEmployeeMerged = pdEmployeeMerged.drop(['department','dept_id','dept_head'], axis=1)

In [0]:
pdEmployeeMerged.head()

avg_monthly_hrs  filed_complaint  ...  marital_status  dept_name
0            246.0              0.0  ...         Married      Sales
1            134.0              0.0  ...       Unmarried      Sales
2            156.0              1.0  ...       Unmarried      Sales
3            256.0              0.0  ...         Married    Support
4            146.0              0.0  ...       Unmarried      Sales

[5 rows x 14 columns]

In [0]:
# Let's convert categorical Text values in Numerical - will be helpful to check correlation

# Marital_Status : Married = 0 Unmarried = 1
pdEmployeeMerged["marriedNum"] = pdEmployeeMerged["marital_status"].astype('category').cat.codes
# gender : Female = 0 Male = 1
pdEmployeeMerged["genderNum"] = pdEmployeeMerged["gender"].astype('category').cat.codes
# Status : Left = 1 Employeed = 0
pdEmployeeMerged["statusNum"] = pdEmployeeMerged["status"].astype('category').cat.codes
# Salary : Let's create dummies
pdEmployeeMerged1 = pd.get_dummies(pdEmployeeMerged,columns=['salary','dept_name'], prefix = ['salary','dept'])

In [0]:
pdEmployeeMerged1.head()
pdEmployeeMerged1 = pdEmployeeMerged1.drop(['status','gender','marital_status'], axis=1)
pdEmployeeMerged1.head()

avg_monthly_hrs  filed_complaint  ...  dept_Support  dept_Temp
0            246.0              0.0  ...             0          0
1            134.0              0.0  ...             0          0
2            156.0              1.0  ...             0          0
3            256.0              0.0  ...             1          0
4            146.0              0.0  ...             0          0

[5 rows x 26 columns]

In [0]:
profile = pandas_profiling.ProfileReport(pdEmployeeMerged1)
profile.to_file(outputfile="pdEmployeeMerged1.html")  

In [0]:
files.download("pdEmployeeMerged1.html")

In [0]:
pdEmployeeMerged2 = pdEmployeeMerged1.loc[pdEmployeeMerged1["statusNum"] == 1]

In [0]:
profile = pandas_profiling.ProfileReport(pdEmployeeMerged2)
profile.to_file(outputfile="pdEmployeeMerged2.html")  